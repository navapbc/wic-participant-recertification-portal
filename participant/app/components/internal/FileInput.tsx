/* Heavily edited from @trussworks/react-uswds FileInput component */
/* Added:
 *  * Internationalization
 *  * Noscript behavior
 * Removed:
 *  * FilePreview components (now controlled by FileUploader)
 */
import React, {
  useState,
  forwardRef,
  useEffect,
  useRef,
  useImperativeHandle,
} from "react";
import classnames from "classnames";
import type { i18nKey } from "~/types";
import { useTranslation } from "react-i18next";
import { Label } from "@trussworks/react-uswds";

export type FileInputProps = {
  id: string;
  name: string;
  disabled?: boolean;
  multiple?: boolean;
  accept?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onDrop?: (e: React.DragEvent) => void;
  emptyKey: i18nKey;
  notEmptyKey: i18nKey;
  emptyAriaKey: i18nKey;
  notEmptyAriaKey: i18nKey;
  fileTypeErrorKey: i18nKey;
  empty?: boolean;
};

export type FileInputRef = {
  input: HTMLInputElement | null;
  files: File[];
};

export const FileInputForwardRef: React.ForwardRefRenderFunction<
  FileInputRef,
  FileInputProps & JSX.IntrinsicElements["input"]
> = (
  {
    name,
    id,
    disabled,
    multiple,
    className,
    accept,
    onChange,
    onDrop,
    emptyKey,
    notEmptyKey,
    emptyAriaKey,
    notEmptyAriaKey,
    fileTypeErrorKey,
    empty = true,
    ...inputProps
  },
  ref
): React.ReactElement => {
  const internalRef = useRef<HTMLInputElement>(null);
  const [isDragging, setIsDragging] = useState(false);
  const [showError, setShowError] = useState(false);
  const [javascriptDisabled, setjavascriptDisabled] = useState(true);
  const [files, setFiles] = useState<File[]>([]);
  const { t } = useTranslation();
  const helpText = empty ? t(emptyKey) : t(notEmptyKey);
  const ariaHelpText = empty ? t(emptyAriaKey) : t(notEmptyAriaKey);
  useImperativeHandle(
    ref,
    () => ({
      input: internalRef.current,
      clearFiles: (): void => setFiles([]),
      files,
    }),
    [files]
  );

  useEffect(() => {
    setjavascriptDisabled(false);
  }, []);
  const fileInputClasses = classnames(
    {
      "usa-file-input": !javascriptDisabled,
      "usa-file-input--disabled": disabled,
    },
    className
  );

  const targetClasses = classnames(
    "usa-file-input__target",
    {
      "usa-file-input--drag": isDragging,
      "has-invalid-file": showError,
    },
    "border-dashed",
    "border-blue",
    "border-05"
  );

  const instructionClasses = classnames("usa-file-input__instructions");

  const preventInvalidFiles = (e: React.DragEvent): void => {
    setShowError(false);

    if (accept) {
      const acceptedTypes = accept.split(",");
      let allFilesAllowed = true;
      for (let i = 0; i < e.dataTransfer.files.length; i += 1) {
        const file = e.dataTransfer.files[parseInt(`${i}`)];
        if (allFilesAllowed) {
          for (let j = 0; j < acceptedTypes.length; j += 1) {
            const fileType = acceptedTypes[parseInt(`${j}`)];
            allFilesAllowed =
              file.name.indexOf(fileType) > 0 ||
              file.type.includes(fileType.replace(/\*/g, ""));
            if (allFilesAllowed) break;
          }
        } else break;
      }

      if (!allFilesAllowed) {
        setFiles([]);
        setShowError(true);
        e.preventDefault();
        e.stopPropagation();
      }
    }
  };

  // Event handlers
  const handleDragOver = (): void => setIsDragging(true);
  const handleDragLeave = (): void => setIsDragging(false);
  const handleDrop = (e: React.DragEvent): void => {
    preventInvalidFiles(e);
    setIsDragging(false);
    if (onDrop) onDrop(e);
  };
  const inputBoxClass = classnames({
    "usa-file-input__box": !javascriptDisabled,
  });
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setShowError(false);

    // Map input FileList to array of Files
    const fileArr = [];
    if (e.target?.files?.length) {
      const fileLength = e.target?.files?.length || 0;

      for (let i = 0; i < fileLength; i++) {
        const file = e.target.files.item(i);
        if (file) fileArr.push(file);
      }
    }
    setFiles(fileArr);

    if (onChange) onChange(e);
  };

  return (
    <div
      data-testid="file-input"
      className={fileInputClasses}
      aria-disabled={disabled}
    >
      <div
        data-testid="file-input-droptarget"
        className={targetClasses}
        onDragOver={handleDragOver}
        onDragLeave={handleDragLeave}
        onDrop={handleDrop}
        hidden={javascriptDisabled}
      >
        <Label
          data-testid="file-input-instructions"
          className={instructionClasses}
          id="file-input-instructions"
          aria-label={ariaHelpText}
          htmlFor={id}
        >
          <span
            className="usa-file-input__choose text-bold"
            hidden={javascriptDisabled}
          >
            {helpText}
          </span>
        </Label>
        <div data-testid="file-input-box" className={inputBoxClass}></div>
        {showError && (
          <div
            data-testid="file-input-error"
            className="usa-file-input__accepted-files-message"
          >
            {t(fileTypeErrorKey)}
          </div>
        )}
        <input
          {...inputProps}
          ref={internalRef}
          type="file"
          data-testid="file-input-input"
          name={name}
          id={id}
          className="usa-file-input__input"
          disabled={disabled}
          onChange={handleChange}
          accept={accept}
          multiple={multiple || javascriptDisabled}
        />
      </div>
    </div>
  );
};

export const FileInput = forwardRef(FileInputForwardRef);
