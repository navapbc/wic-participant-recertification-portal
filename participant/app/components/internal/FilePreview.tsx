/* Lifted and edited from @trussworks/react-uswds FilePreview internal only component
 * This component (in the source) gets displayed as part of the @trussworks/react-uswds
 * FileInput component after a file (or files) are selected
 * Added:
 * * Localization strings for the text
 * * Button that triggers a callback (currently used as "Remove File")
 * * clickHandler for the callback
 * * Props definition
 * Changed:
 * * Replaced use of fileReaderRef with URL.createObjectURL
 * * Minor className changes for styling
 */
import React, { useEffect, useState } from "react";
import type { ReactElement } from "react";
import classNames from "classnames";
import type { i18nKey } from "~/types";
import { useTranslation } from "react-i18next";
/** Moving the SPACER_GIF definition here instead of the constants.ts file,
 * as webpack was exporting that entire file, including use of the File
 * WebAPI; this was causing server-side site generators to break (#1250). */

const SPACER_GIF =
  "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";

export type FilePreviewProps = {
  imageId: string;
  file: File;
  clickHandler: Function;
  removeFileKey: i18nKey;
  selectedKey: i18nKey;
  altTextKey: i18nKey;
};

export const FilePreview = (props: FilePreviewProps): ReactElement => {
  const {
    imageId,
    file,
    clickHandler,
    removeFileKey,
    selectedKey,
    altTextKey,
  } = props;

  const [isLoading, setIsLoading] = useState(true);
  const [previewSrc, setPreviewSrc] = useState(SPACER_GIF);
  const [showGenericPreview, setShowGenericPreview] = useState(false);
  const { t } = useTranslation();

  useEffect(() => {
    const imgSrc = URL.createObjectURL(file);
    setIsLoading(false);
    setPreviewSrc(imgSrc);
  }, [file]);

  const { name } = file;

  const onImageError = (): void => {
    setPreviewSrc(SPACER_GIF);
    setShowGenericPreview(true);
  };

  const isPDF = name.indexOf(".pdf") > 0;
  const isWord = name.indexOf(".doc") > 0 || name.indexOf(".pages") > 0;
  const isVideo = name.indexOf(".mov") > 0 || name.indexOf(".mp4") > 0;
  const isExcel = name.indexOf(".xls") > 0 || name.indexOf(".numbers") > 0;
  const isGeneric = !isPDF && !isWord && !isVideo && !isExcel;

  const imageClasses = classNames("usa-file-input__preview-image", {
    "is-loading": isLoading,
    "usa-file-input__preview-image--pdf": showGenericPreview && isPDF,
    "usa-file-input__preview-image--word": showGenericPreview && isWord,
    "usa-file-input__preview-image--video": showGenericPreview && isVideo,
    "usa-file-input__preview-image--excel": showGenericPreview && isExcel,
    "usa-file-input__preview-image--generic": showGenericPreview && isGeneric,
  });
  const altText = `${t(altTextKey)} ${name}`;
  return (
    <div className="margin-bottom-4 margin-top-2">
      <div className="usa-file-input">
        <div className="usa-file-input__target">
          <div className="usa-file-input__preview-heading">
            <div>{t(selectedKey)}</div>
            <button
              type="button"
              className="text-secondary-vivid usa-button--unstyled"
              onClick={() => clickHandler(name)}
            >
              {t(removeFileKey)}
            </button>
          </div>
          <div className="usa-file-input__preview" aria-hidden="true">
            <img
              id={imageId}
              src={previewSrc}
              alt={altText}
              className={imageClasses}
              onError={onImageError}
            />
            {name}
          </div>
          <div className="usa-file-input__box" />
        </div>
      </div>
    </div>
  );
};
