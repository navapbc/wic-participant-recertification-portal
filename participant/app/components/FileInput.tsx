import { FormGroup, Label } from "@trussworks/react-uswds";
import { FilePreview } from "app/components/internal/FilePreview";
import { FileInputControl } from "app/components/internal/FileInputControl";
import React, {
  useState,
  useRef,
  useImperativeHandle,
  useEffect,
  forwardRef,
} from "react";
import { i18nKey } from "~/types";
import { Trans, useTranslation } from "react-i18next";

export type FileInputProps = {
  className?: string;
  id: string;
  name: string;
  labelKey: i18nKey;
  accept?: string;
  required?: boolean;
  maxFileSizeInBytes?: number;
  maxFileCount?: number;
};

export type FileState = {
  [key: string]: File;
};

export type FileInputRef = {
  input: HTMLInputElement | null;
  files: File[];
};

export type FileError = {
  errorType?: "count" | "size" | "type";
};

export const FileInputForwardRef: React.ForwardRefRenderFunction<
  FileInputRef,
  FileInputProps & JSX.IntrinsicElements["input"]
> = (
  {
    className,
    id,
    name,
    labelKey,
    accept,
    required,
    maxFileSizeInBytes = 26_214_400,
    maxFileCount = 25,
    ...inputProps
  },
  ref
): React.ReactElement => {
  const { t } = useTranslation();
  const [files, setFiles] = useState({} as FileState);
  const [previews, setPreviews] = useState(<></>);
  const fileInputRef = useRef(null);
  const documentString = t(`${labelKey}.document`);
  const [errorMessage, setErrorMessage] = useState("");

  const ifNeeded = t(`${labelKey}.ifNeeded`);
  const currentDocumentNumber = Object.keys(files).length;

  useImperativeHandle(
    ref,
    () => ({
      input: fileInputRef.current,
      files: Object.keys(files).map((key) => files[key]),
    }),
    [files]
  );
  const makeSafeForID = (name: string): string => {
    return name.replace(/[^a-z0-9]/g, function replaceName(s) {
      const c = s.charCodeAt(0);
      if (c === 32) return "-";
      if (c >= 65 && c <= 90) return `img_${s.toLowerCase()}`;
      return `__${c.toString(16).slice(-4)}`;
    });
  };

  const renderPreviews = () => {
    setPreviews(
      <>
        {Object.keys(files).map((fileName, index) => {
          let file = files[fileName];
          const imageId = makeSafeForID(fileName);
          return (
            <div key={`document-${index + 1}`}>
              <span className="font-sans-lg">{`${documentString} ${
                index + 1
              }`}</span>
              <FilePreview
                imageId={`preview-${index}`}
                file={file}
                clickHandler={removeFile}
                removeFileKey={`${labelKey}.removeFile`}
                selectedKey={`${labelKey}.selected`}
                altTextKey={`${labelKey}.altText`}
              />
            </div>
          );
        })}
      </>
    );
  };
  useEffect(renderPreviews, [files]);
  const fileTypeCheck = (newFile: File): boolean => {
    if (!accept || accept == "*") {
      return true;
    }
    const acceptedTypes = accept.split(",");
    let fileTypeAllowed = true;
    if (fileTypeAllowed) {
      for (let j = 0; j < acceptedTypes.length; j += 1) {
        const fileType = acceptedTypes[parseInt(`${j}`)];
        fileTypeAllowed =
          newFile.name.indexOf(fileType) > 0 ||
          newFile.type.includes(fileType.replace(/\*/g, ""));
        if (fileTypeAllowed) break;
      }
    }
    return fileTypeAllowed;
  };
  const addNewFiles = (newFiles: FileList) => {
    setErrorMessage("");

    if (
      Object.keys(files).length + Object.keys(newFiles).length <=
      maxFileCount
    ) {
      for (let file of newFiles) {
        if (file.size <= maxFileSizeInBytes) {
          if (fileTypeCheck(file)) {
            files[file.name] = file;
          } else {
            setErrorMessage(`${labelKey}.fileTypeError`);
          }
        } else {
          setErrorMessage(`${labelKey}.fileSizeError`);
        }
      }
    } else {
      setErrorMessage(`${labelKey}.fileCountError`);
    }
    return { ...files };
  };
  const handleNewFileUpload = (e: React.ChangeEvent<HTMLInputElement>) => {
    setErrorMessage("");
    const { files: newFiles } = e.target;
    if (newFiles?.length) {
      let updatedFiles = addNewFiles(newFiles);
      setFiles(updatedFiles);
    }
  };
  const removeFile = (fileName: string) => {
    setErrorMessage("");
    delete files[fileName];
    setFiles({ ...files });
  };

  const currentDocumentString =
    currentDocumentNumber > 0
      ? `${documentString} ${currentDocumentNumber + 1} ${ifNeeded}`
      : `${documentString} ${currentDocumentNumber + 1}`;
  return (
    <FormGroup>
      {previews}
      <Label htmlFor="file-input-async" className="font-sans-lg">
        {currentDocumentString}
      </Label>
      {errorMessage && (
        <div data-testid="file-input-error" className="file-input-error">
          <Trans i18nKey={errorMessage} />
        </div>
      )}
      <FileInputControl
        {...inputProps}
        ref={fileInputRef}
        id={id}
        name={name}
        accept={accept}
        onChange={handleNewFileUpload}
        onDrop={function noRefCheck() {}}
        value=""
        title=""
        draggable={true}
        emptyKey={`${labelKey}.noFiles`}
        notEmptyKey={`${labelKey}.additionalFiles`}
        fileTypeErrorKey={`${labelKey}.fileTypeError`}
        empty={currentDocumentNumber == 0}
      />
    </FormGroup>
  );
};

export const FileInput = forwardRef(FileInputForwardRef);
