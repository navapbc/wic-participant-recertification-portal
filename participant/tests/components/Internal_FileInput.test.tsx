import { FileInput } from "app/components/internal/FileInput";
import type { FileInputProps } from "app/components/internal/FileInput";
import { renderWithRouter } from "tests/helpers/setup";

const defaultProps: FileInputProps = {
  emptyKey: "test:fileuploader.noFiles",
  notEmptyKey: "test:fileuploader.additionalFiles",
  emptyAriaKey: "test:fileuploader.noFilesAria",
  notEmptyAriaKey: "test:fileuploader.additionalFilesAria",
  fileTypeErrorKey: "test:fileuploader.fileTypeError",
  name: "fileinput-test",
  id: "fileinput-test",
};

it("renders a file input correctly", () => {
  const { container } = renderWithRouter(<FileInput {...defaultProps} />);
  expect(container).toMatchSnapshot();
});

it("renders a file input with the 'not empty' text correctly", () => {
  const { container } = renderWithRouter(
    <FileInput {...defaultProps} empty={false} />
  );
  expect(container).toMatchSnapshot();
});
