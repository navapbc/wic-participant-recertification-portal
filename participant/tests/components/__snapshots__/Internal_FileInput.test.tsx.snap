// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a file input correctly 1`] = `
<div>
  <div
    class="usa-file-input"
    data-testid="file-input"
  >
    <div
      class="usa-file-input__target border-dashed border-blue border-05"
      data-testid="file-input-droptarget"
    >
      <label
        class="usa-label usa-file-input__instructions"
        data-testid="label"
        for="fileinput-test"
        id="file-input-instructions"
      >
        <span
          class="usa-file-input__choose text-bold"
        >
          Choose or drag a file to upload
        </span>
      </label>
      <div
        class="usa-file-input__box"
        data-testid="file-input-box"
      />
      <input
        class="usa-file-input__input"
        data-testid="file-input-input"
        id="fileinput-test"
        name="fileinput-test"
        type="file"
      />
    </div>
  </div>
</div>
`;

exports[`renders a file input with the 'not empty' text correctly 1`] = `
<div>
  <div
    class="usa-file-input"
    data-testid="file-input"
  >
    <div
      class="usa-file-input__target border-dashed border-blue border-05"
      data-testid="file-input-droptarget"
    >
      <label
        class="usa-label usa-file-input__instructions"
        data-testid="label"
        for="fileinput-test"
        id="file-input-instructions"
      >
        <span
          class="usa-file-input__choose text-bold"
        >
          Choose or drag another file to upload
        </span>
      </label>
      <div
        class="usa-file-input__box"
        data-testid="file-input-box"
      />
      <input
        class="usa-file-input__input"
        data-testid="file-input-input"
        id="fileinput-test"
        name="fileinput-test"
        type="file"
      />
    </div>
  </div>
</div>
`;
